book.md

```diff
@@ -398,15 +398,76 @@ cache, without the processes having rights to read or write to the same memory
+the CPU can continue executing quickly; if not, dependent operations will have
+to wait until the data returns from the much more slowly main memory.
+Typical access times is 3 to 5 CPU cycles for the fastest cache on a CPU versus
+hundreds of cycles for a main memory access.\index{memory access time}
+
+Most systems have multiple level of cache\index{multi-level cache}, each with a
+different trade-off between cache size\index{cache size} and access
+time\index{cache access time}. Some typical characteristics might be:
+
+* A L1 (level 1) cache, 32kB in size, with an access time of 4 cycles.
+* A L2 cache, 256Kb in size, with an access time of 10 cycles.
+* A L3 cache, 16MB in size, with an access time of 40 cycles.
+* Main memory, gigabytes in size, with an access time of 100 cycles.
+
+When the data was not present in the cache, it typically is stored in the cache
+after it's fetched from a slower cache level or main memory. This is often a
+good decision to make as there's a high likelihood the same address will be
+accessed by the program soon after. This high likelihood is known as the
+[principle of locality](https://en.wikipedia.org/wiki/Locality_of_reference)\index{principle
+of locality}\index{locality of reference}.
+
+Data is stored and transferred between cache levels in blocks of aligned memory.
+Such a block is called a \index{cache block} or \index{cache line}. Typical
+sizes are 32, 64 or 128 bytes per cache line.
+
+When data that wasn't previously in the cache needs to be stored in the cache,
+most of the time, room has to be made for it be removing, or
+evicting\index{cache eviction}, some other address/data from it. How that choice
+gets made is decided by the
+[cache replacement policy](https://en.wikipedia.org/wiki/Cache_replacement_policies)\index{cache
+replacement policy}. Popular replacement algorithms are Least Recently Used
+(LRU)\index{LRU replacement policy}, Random\index{random replacement policy} and
+pseudo-LRU\index{pseudo-LRU replacement policy}. As the names suggest, LRU
+evicts the cache line that is least recently used; random picks a random cache
+line; and pseudo-LRU approximates choosing the least recently used line.
+
+If an address can be stored in all locations available in the cache, the cache
+is fully-associative\index{fully-associative cache}. Most caches are however not
+fully-associative, as it's too costly to implement. Instead, most caches are
+set-associative\index{set-associative cache}. In a set-associate cache, a
+specific address can only be stored in a few locations. If an address can
+potentially be stored in 2 locations, the cache is said to be 2-way set
+associative. If it can be stored in one of at most 4 locations, it's called
+4-way set associative, and so on. When an address can only be stored in one
+location in the call, the cache is said to be direct mapped\index{direct mapped
+cache}, rather than 1-way set associative. Typical values of set associativity
```

Comment:
"Typical values of set associativity..." : I find this sentence a little awkward grammatically. But also a direct mapped cache is not set-associative. Perhaps this could be: "Typical cache organizations are..." WDYT?

Reply:
Yeah - I went for "Typical organizations are..."